frame/support/src/traits/hooks.rs

```diff
@@ -135,17 +144,25 @@ pub trait OnRuntimeUpgrade {
 
 	/// Execute some pre-checks prior to a runtime upgrade.
 	///
+	/// Return the state digest which will be passed to `post_upgrade` after upgrading for
+	/// post-check. NOTE: if this function is not implemented, `Self::PreStateDigest` should be set
+	/// to `()`.
+	///
 	/// This hook is never meant to be executed on-chain but is meant to be used by testing tools.
-	#[cfg(feature = "try-runtime")]
-	fn pre_upgrade() -> Result<(), &'static str> {
-		Ok(())
-	}
+	/// TODO: add #[cfg(feature = "try-runtime")], which required changing a lot of `Cargo.toml`
+	/// files to recorrect features dependencies
+	fn pre_upgrade() -> Result<Self::PreStateDigest, &'static str>;
```

Comment:
I had tried to provide a default implement here by `PreStateDigest: Default` and return `Ok(Self:: PreStateDigest::default())`. But it failed to implement `OnRuntimeUpgrade for Tuple` because it has `PreStateDigest: (P1, P2, ..)` and rust only implement `Default` for tuples up to 12 elements, also https://github.com/paritytech/substrate/pull/11813 had changed the nested tuple to a flatten one.

Reply:
Oops, that is one issue that we may have intorduced with flatten tuples.